aspnetcore/client-side/spa-services.md

```diff
@@ -0,0 +1,328 @@
+
+## Server-side prerendering
+
+A universal (also known as isomorphic) application is a JavaScript application capable of running both on the server and the client. Angular, React, and other popular frameworks provide a universal platform for this application development style. The idea is to first render the framework components on the server via Node.js and then delegate further execution to the client.
+
+SpaServices' ASP.NET Core [Tag Helpers](xref:mvc/views/tag-helpers/intro) simplify the implementation of server-side prerendering by invoking the JavaScript functions on the server for you.
+
+### Prerequisites
+
+Install the following prerequisites:
+1. [Microsoft.AspNetCore.SpaServices](http://www.nuget.org/packages/Microsoft.AspNetCore.SpaServices/) NuGet package
+1. [aspnet-prerendering](https://www.npmjs.com/package/aspnet-prerendering) npm package:
+
+    ```console
+    npm i -S aspnet-prerendering
+    ```
+
+### Configuration
+
+The Tag Helpers are made discoverable via namespace registration in the project's *_ViewImports.cshtml* file:
+
+[!code-csharp[Main](../client-side/spa-services/sample/SpaServicesSampleApp/Views/_ViewImports.cshtml?highlight=3)]
+
+These Tag Helpers abstract away the intricacies of communicating directly with low-level APIs by leveraging an HTML-like syntax inside the Razor view:
+
+[!code-html[Main](../client-side/spa-services/sample/SpaServicesSampleApp/Views/Home/Index.cshtml?range=5)]
+
+> [!TIP]
+> Microsoft's **[Razor Language Services](https://marketplace.visualstudio.com/items?itemName=ms-madsk.RazorLanguageServices)** extension improves Visual Studio 2017's Tag Helpers development experience by adding context-aware IntelliSense and syntax highlighting:
+>
+> ![Tag Helpers intellisense](../client-side/spa-services/_static/tag_helper_intellisense.png)
+
+### The `asp-prerender-module` Tag Helper
+
+The `asp-prerender-module` Tag Helper, used in the preceding code example, executes *ClientApp/dist/main-server.js* on the server via Node.js. For clarity's sake, *main-server.js* file is an artifact of the [Webpack](http://webpack.github.io/) build process' TypeScript-to-JavaScript transpilation task. Webpack defines an entry point alias of `main-server`; and, traversal of the dependency graph for this alias begins at the *ClientApp/boot-server.ts* file:
+
+[!code-javascript[Main](../client-side/spa-services/sample/SpaServicesSampleApp/webpack.config.js?range=53)]
+
+In the following Angular example, the *ClientApp/boot-server.ts* file utilizes the `createServerRenderer` function and `RenderResult` type of the `aspnet-prerendering` npm package to configure server rendering via Node.js. The HTML markup destined for server-side rendering is passed to a `resolve` function call, which is wrapped in a JavaScript `Promise` object of type `RenderResult`. The `Promise` object's significance is that it asynchronously supplies the HTML markup to the page for injection in the DOM's placeholder element.
```

Comment:
We really like to avoid english versions of code. Would the following work? Do you really need all that type info:

The HTML markup destined for server-side rendering is passed to a `resolve` function call, which is wrapped in a JavaScript `Promise` object ~of type `RenderResult`.~

or even better
The HTML markup destined for server-side rendering is passed to a `resolve` function call, ~which is wrapped in a JavaScript `Promise` object of type `RenderResult`.~